heimlig/src/hsm/workers/ecc_worker.rs

```diff
@@ -70,94 +125,337 @@ impl<
+        request_id: RequestId,
+        public_key: &[u8],
+        message: &[u8],
+        prehashed: bool,
+        signature: &[u8],
+    ) -> Response<'data> {
+        let result = match public_key.len() {
+            crypto::ecdsa::NIST_P256_PUBLIC_KEY_SIZE => {
+                if prehashed {
+                    nist_p256_verify_prehashed(public_key, message, signature)
+                } else {
+                    nist_p256_verify(public_key, message, signature)
+                }
+            }
+            crypto::ecdsa::NIST_P384_PUBLIC_KEY_SIZE => {
+                if prehashed {
+                    nist_p384_verify_prehashed(public_key, message, signature)
+                } else {
+                    nist_p384_verify(public_key, message, signature)
+                }
+            }
+            _ => {
+                return Response::Error {
+                    client_id,
+                    request_id,
+                    error: Error::Crypto(crypto::Error::InvalidPrivateKey),
+                };
+            }
+        };
+
+        match result {
+            Err(crypto::Error::InvalidSignature) => Response::Verify {
+                client_id,
+                request_id,
+                verified: false,
+            },
+            Err(e) => Response::Error {
+                client_id,
+                request_id,
+                error: Error::Crypto(e),
+            },
+            Ok(_) => Response::Verify {
+                client_id,
+                request_id,
+                verified: true,
+            },
+        }
+    }
+
+    async fn export_private_key_and_key_info<'a>(
+        &mut self,
+        key_id: KeyId,
+        key_buffer: &'a mut [u8],
+    ) -> Result<(&'a [u8], KeyInfo), keystore::Error> {
+        // Lock keystore only once
+        let locked_key_store = self.key_store.lock().await;
+
+        Ok((
+            locked_key_store.export_private_key_unchecked(key_id, key_buffer)?,
+            locked_key_store.get_key_info(key_id)?,
+        ))
+    }
+
+    async fn export_public_key_and_key_info<'a>(
+        &mut self,
+        key_id: KeyId,
+        key_buffer: &'a mut [u8],
+    ) -> Result<(&'a [u8], KeyInfo), keystore::Error> {
+        // Lock keystore only once
+        let locked_key_store = self.key_store.lock().await;
+
+        Ok((
+            locked_key_store.export_public_key(key_id, key_buffer)?,
+            locked_key_store.get_key_info(key_id)?,
+        ))
+    }
+}
+
+fn move_key_pair<'a, const N: usize, const M: usize>(
+    mut private_key: [u8; N],
+    mut public_key: [u8; M],
+    private_key_bytes: &'a mut [u8],
+    public_key_bytes: &'a mut [u8],
+) -> (&'a [u8], &'a [u8]) {
+    let private_key_result = &mut private_key_bytes[..private_key.len()];
+    private_key_result.copy_from_slice(&private_key);
+    let public_key_result = &mut public_key_bytes[..public_key.len()];
+    public_key_result.copy_from_slice(&public_key);
+
+    private_key.zeroize();
```

Comment:
As it is used now, the mutable input parameters are wrapped into `Zeroize` on the caller site anyway so we zeroize a second time here, right? If so, why do we have `move_key_pair` and not just copy from key to key on the caller site and let `Zeroize` do the cleanup?

Reply:
As discussed offline, these explicit `zeroize()` calls are used for temporary array variables.